src/aspire/source/__init__.py

```diff
@@ -363,12 +359,10 @@ def whiten(self, noise_filter):
         whiten_filter = PowerFilter(noise_filter, power=-0.5)
 
         logger.info('Transforming all CTF Filters into Multiplicative Filters')
-        unique_filters = set(self.filters)
-        for f in unique_filters:
-            f_new = copy(f)
-            f.__class__ = MultiplicativeFilter
-            f.__init__(f_new, whiten_filter)
-
+        for filt in self.unique_filters:
```

Comment:
One of the main reasons we wanted to move away from the previous list of filter pointers was to avoid this kind of in-place modification of the filter objects. Why are we still doing it?

Reply:
We assigned this immutable problem for another PR. I can update it if we would like to do it in this PR.

Reply:
This has nothing to do with immutability. Since the `filter_indices` point to an index in the `unique_filters` list, we can just replace the filter at that index without having to do any in-place modifications of the filters as is done here.

Reply:
Thanks. Statements about go through the unique filters only instead of going through all filters in filter_indices. The immutable updates has been done locally.  The statements were replaced as `self.unique_filters = [MultiplicativeFilter(f, whiten_filter) for f in self.unique_filters]`. Can I do this in new PR so that no need to rearrange commits?

Reply:
> The statements were replaced as `self.unique_filters = [MultiplicativeFilter(f, whiten_filter) for f in self.unique_filters]`. 

Yes. That is the correct way to do it now that we no longer depend on pointers.

> Can I do this in new PR so that no need to rearrange commits? 

I don't understand why we can't do this now. This was one of the reasons we're switching the way we represent filters in the first place.

Reply:
Done in commit 0d7aeed.